In [1]:
#*******************************************************************************************
 #
 #  File Name:  CreditRiskClassification.ipynb
 #
 #  File Description:
 #      This interactive Python notebook, CreditRiskClassification.ipynb, reads a csv file,
 #      LendingData.csv, in the Resources folder,  and uses Python and supervised 
 #      learning methods to use a dataset of historical lending activity from a peer-to-peer
 #      lending services company to build a model that can identify the creditworthiness of 
 #      borrowers. 
 #
 #      I reinstalled the scikit-learn module using the following commands to allow the
 #      RandomOverSampler function to work:
 #
 #      pip3 uninstall scikit-learn
 #      pip3 install scikit-learn==1.2.2
 #
 #
 #  Date            Description                             Programmer
 #  ----------      ------------------------------------    ------------------
 #  11/25/2023      Initial Development                     N. James George
 #
 #******************************************************************************************/

import PyConstants as constant
import PyFunctions as function
import PyLogConstants as log_constant
import PyLogFunctions as log_function
import PyLogSubRoutines as log_subroutine
import PySubRoutines as subroutine

import numpy as np
import pandas as pd

from pathlib import Path

from imblearn.over_sampling import RandomOverSampler

from sklearn.linear_model import LogisticRegression
from sklearn.metrics import balanced_accuracy_score, confusion_matrix, classification_report
from sklearn.model_selection import train_test_split

In [2]:
CONSTANT_LOCAL_FILE_NAME \
    = 'CreditRiskClassification.ipynb'

CONSTANT_LENDING_DATA_CSV_FILE_PATH \
    = './Resources/LendingData.csv'

log_subroutine \
    .SetLogMode \
        (False)

log_subroutine \
    .SetDebugMode \
        (False)

log_subroutine \
    .SetImageMode \
        (False)

log_subroutine \
    .BeginProgramExecution \
        ('CreditRiskClassification')

# <br> **Section 1: Split the Data into Training and Testing Sets**

## **1.1: Read the CSV data from the `Resources` folder into a Pandas DataFrame**

### **Read the CSV data from the `Resources` folder**

In [3]:
lendingDataFrame \
    = function. \
        ReturnCSVFileAsDataFrame \
            (CONSTANT_LENDING_DATA_CSV_FILE_PATH)


log_function \
    .DebugReturnObjectWriteObject \
        (lendingDataFrame)

### **Display Lending DataFrame**

In [4]:
captionString \
    = 'Table 1.2.1: Lending Data Table'

currentStylerObject \
    = function \
        .ReturnStylerObjectStandardFormat \
            (lendingDataFrame.head(12),
             captionString) \
        .format({'loan_size': '${:,.0f}',
                 'interest_rate': '{:.2f}%',
                 'borrower_income': '${:,.0f}',
                 'total_debt': '${:,.0f}'})

log_function \
    .ReturnStylerObjectSavePNGImage \
        (currentStylerObject,
         captionString)

## **1.2: Create the labels set (`y`)  from the “loan_status” column, and then create the features (`X`) DataFrame from the remaining columns.**

### **Separate the y variable, the labels**

In [5]:
ySeries \
    = lendingDataFrame \
        ['loan_status']

log_function \
    .DebugReturnObjectWriteObject \
        (ySeries)

### **Separate the X variable, the features**

In [6]:
xDataFrame \
    = lendingDataFrame \
        .drop \
            (columns = 'loan_status',
             axis = 1)

log_function \
    .DebugReturnObjectWriteObject \
        (xDataFrame)

### **Review the Y Variable Series**

In [7]:
captionString \
    = 'Table 1.3.1: Y Variable Series'

currentStylerObject \
    = function \
        .ReturnStylerObjectStandardFormat \
            (ySeries.to_frame().head(12),
             captionString)

log_function \
    .ReturnStylerObjectSavePNGImage \
        (currentStylerObject,
         captionString)

### **Review the X Variable DataFrame**

In [8]:
captionString \
    = 'Table 1.3.2: X Variable DataFrame'

currentStylerObject \
    = function \
        .ReturnStylerObjectStandardFormat \
            (xDataFrame.head(12),
             captionString) \
        .format({'loan_size': '${:,.0f}',
                 'interest_rate': '{:.2f}%',
                 'borrower_income': '${:,.0f}',
                 'total_debt': '${:,.0f}'})

log_function \
    .ReturnStylerObjectSavePNGImage \
        (currentStylerObject,
         captionString)

## **1.3: Check the balance of the labels variable (`y`) by using the `value_counts` function.**

In [9]:
ySeries.value_counts()

loan_status
0    75036
1     2500
Name: count, dtype: int64

## **1.4: Split the data into training and testing datasets by using `train_test_split`.**

In [10]:
xTrainDataFrame, \
xTestDataFrame, \
yTrainSeries, \
yTestSeries \
    = train_test_split \
        (xDataFrame, 
         ySeries, 
         random_state = 1)

In [11]:
log_function \
    .DebugReturnObjectWriteObject \
        (xTrainDataFrame)

In [12]:
log_function \
    .DebugReturnObjectWriteObject \
        (xTestDataFrame)

In [13]:
log_function \
    .DebugReturnObjectWriteObject \
        (yTrainSeries)

In [14]:
log_function \
    .DebugReturnObjectWriteObject \
        (yTestSeries)

# <br> **Section 2: Create a Logistic Regression Model with the Original Data**

## **2.1: Fit a logistic regression model by using the training data.**

In [15]:
# This line of code assigns a random_state parameter of 1 to the model
# and fits the model using training data.
modelLogisticRegression \
    = LogisticRegression \
        (random_state = 1) \
            .fit \
                (xTrainDataFrame, 
                 yTrainSeries)

log_function \
    .DebugReturnObjectWriteObject \
        (modelLogisticRegression)

## **2.2: Save the predictions on the testing data labels by using the testing feature data and the fitted model.**

In [16]:
predictionsNumpyArray \
    = modelLogisticRegression \
        .predict \
            (xTestDataFrame)

log_function \
    .DebugReturnObjectWriteObject \
        (predictionsNumpyArray)

In [17]:
predictionDictionary \
    = {'Prediction': predictionsNumpyArray, 
       'Actual': yTestSeries}

predictionDataFrame \
    = pd.DataFrame \
        (predictionDictionary)

log_function \
    .DebugReturnObjectWriteObject \
        (predictionDataFrame)

In [18]:
captionString \
    = 'Table 2.2.1: Testing Data Label Predictions (Original)'

currentStylerObject \
    = function \
        .ReturnStylerObjectStandardFormat \
            (predictionDataFrame.head(12),
             captionString)

log_function \
    .ReturnStylerObjectSavePNGImage \
        (currentStylerObject,
         captionString)

## **2.3: Evaluate the model’s performance**

In [19]:
def ModelPerformanceEvaluatorFunction \
        (yTestDataSeries, 
         predictionsNumpyArray, 
         originalFlagBoolean = True):

    if originalFlagBoolean == True:
        
        dataTypeString ='Original'
    
    else:
        
        dataTypeString = 'Resampled'
        
        
    # Ths line of code calculates the accuracy score of the model.
    accuracyScoreFloat \
        = balanced_accuracy_score \
            (yTestSeries, 
             predictionsNumpyArray)
    
    # These lines of code generate a confusion matrix DataFrame for display.
    confusionMatrixNumpyArray \
        = confusion_matrix \
            (yTestSeries , 
             predictionsNumpyArray)
    
    indexStringList \
        = ['Actual Healthy', 
           'Actual High-Risk']
    
    columnStringList \
        = ['Predicted Healthy', 
           'Predicted High-Risk']
    
    confusionMatrixDataFrame \
        = pd.DataFrame \
            (confusionMatrixNumpyArray,
             index = indexStringList,
             columns = columnStringList)
    
    # This line of code creates the classification report.
    classificationReportString \
        = classification_report \
            (yTestSeries, 
             predictionsNumpyArray, 
             target_names = ['healthy', 'high risk'])
    
    
    log_subroutine \
        .PrintAndLogWriteText \
            ('\033[1m' 
             + f'LOGISTIC REGRESSION MODEL ({dataTypeString})\n'
             + '\033[0m\n'
             + '1) '
             + '\033[1m'
             + 'Accuracy Score: '
             + '\033[0m'
             + f'{round(accuracyScoreFloat*100, 1)}%\n\n'
             + '2) '
             + '\033[1m'
             + 'Confusion Matrix:\n'
             + '\033[0m\n'
             + f'{confusionMatrixDataFrame}\n\n'
             + f'3) '
             + '\033[1m'
             + 'Classification Report:\n'
             + '\033[0m\n'
             + f'{classificationReportString}\n')
    
    return \
        accuracyScoreFloat, \
        confusionMatrixDataFrame, \
        classificationReportString

In [20]:
accuracyScoreOriginalFloat, \
confusionMatrixOriginalDataFrame, \
classificationReportOriginalString \
    = ModelPerformanceEvaluatorFunction \
        (yTestSeries,
         predictionsNumpyArray)

LOGISTIC REGRESSION MODEL (Original)

1) Accuracy Score: 95.2%

2) Confusion Matrix:

                  Predicted Healthy  Predicted High-Risk
Actual Healthy                18663                  102
Actual High-Risk                 56                  563

3) Classification Report:

              precision    recall  f1-score   support

     healthy       1.00      0.99      1.00     18765
   high risk       0.85      0.91      0.88       619

    accuracy                           0.99     19384
   macro avg       0.92      0.95      0.94     19384
weighted avg       0.99      0.99      0.99     19384




## **Question:** 
### How well does the logistic regression model predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

## **Answer:**
### This logistic regression model does an excellent job predicting healthy loans and correctly identifies healthy loans with a negligible number of false positives or negatives, an accuracy score of 95.2%, a precision score of 100%, and a recall score of 99%.  Nevertheless, this model less accurately predicts high-risk loans with 102 false positives and 56 false negatives leading to a precision score of 85% and a recall score of 91%. The model's comparatively inadequate performance in predicting high-risk loans is concerning and necessitates further optimization with resampling techniques or additonal high-risk loan data.

# <br> **Section 3: Predict a Logistic Regression Model with Resampled Training Data**

## **3.1: Use the `RandomOverSampler` module from the imbalanced-learn library to resample the data. Be sure to confirm that the labels have an equal number of data points.**

### **Instantiate the Random Oversampler Model**

In [21]:
xResampledDataFrame, \
yResampledSeries \
    = RandomOverSampler \
        (random_state = 1) \
            .fit_resample \
                (xTrainDataFrame, 
                 yTrainSeries)

In [22]:
log_function \
    .DebugReturnObjectWriteObject \
        (xResampledDataFrame)

In [23]:
log_function \
    .DebugReturnObjectWriteObject \
        (yResampledSeries)

### **Count the distinct values of the resampled labels data**

In [24]:
yResampledSeries.value_counts()

loan_status
0    56271
1    56271
Name: count, dtype: int64

## **3.2: Use the `LogisticRegression` classifier and the resampled data to fit the model and make predictions.**

### **Fit a logistic regression model by using the training data.**

In [25]:
# This line of code assigns a random_state parameter of 1 to the model
# and fits the model using training data.
modelResampledLogisticRegression \
    = LogisticRegression \
        (random_state = 1) \
            .fit \
                (xResampledDataFrame, 
                 yResampledSeries)

log_function \
    .DebugReturnObjectWriteObject \
        (modelResampledLogisticRegression)

### **Save the predictions on the testing data labels by using the testing feature data and the fitted model.**

In [27]:
predictionsResampledNumpyArray \
    = modelResampledLogisticRegression \
        .predict \
            (xTestDataFrame)

log_function \
    .DebugReturnObjectWriteObject \
        (predictionsResampledNumpyArray)

In [28]:
predictionResampledDictionary \
    = {'Prediction': predictionsResampledNumpyArray, 
       'Actual': yTestSeries}

predictionResampledDataFrame \
    = pd.DataFrame \
        (predictionResampledDictionary) \
            .reset_index \
                (drop = True)

log_function \
    .DebugReturnObjectWriteObject \
        (predictionResampledDataFrame)

In [29]:
captionString \
    = 'Table 2.2.1: Testing Data Label Predictions (Resampled)'

currentStylerObject \
    = function \
        .ReturnStylerObjectStandardFormat \
            (predictionResampledDataFrame.head(12),
             captionString)

log_function \
    .ReturnStylerObjectSavePNGImage \
        (currentStylerObject,
         captionString)

## **3.3: Evaluate the model’s performance with resampled data.**

In [30]:
accuracyScoreOriginalFloat, \
confusionMatrixOriginalDataFrame, \
classificationReportOriginalString \
    = ModelPerformanceEvaluatorFunction \
        (yTestSeries,
         predictionsResampledNumpyArray,
         False)

LOGISTIC REGRESSION MODEL (Resampled)

1) Accuracy Score: 99.4%

2) Confusion Matrix:

                  Predicted Healthy  Predicted High-Risk
Actual Healthy                18649                  116
Actual High-Risk                  4                  615

3) Classification Report:

              precision    recall  f1-score   support

     healthy       1.00      0.99      1.00     18765
   high risk       0.84      0.99      0.91       619

    accuracy                           0.99     19384
   macro avg       0.92      0.99      0.95     19384
weighted avg       0.99      0.99      0.99     19384




## **Question:** 
### How well does the logistic regression model, fit with oversampled data, predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

## **Answer:** 
### 




We can see that the logistic regression model fits well here too with an accuracy score of 99%. Getting into the details, we generally had ideal precision, recall, and f1 for the healthy loan to start with, and now we see those same percentages carry over with the oversampled data. The high-risk loan labels experienced a general increase in percent, except for the precision, which decreased by 1%. The recall and f1 score on the other hand experienced an 8% and 3% increase, respectively.

For the second model, the balanced accuracy score in the f1-score column of the classification report is the same as the accuracy score, 99%, because the data set has an equal number in each class.

In terms of accuracy, the second model trained with resampled data outperforms the first model. This indicates that the ROS technique improved the model's ability to predict the class labels of the test set.
Looking at the confusion matrices, we can see that both models are very good at predicting healthy loans, with a very small number of false positives (116 and 102) for the ROS model and the original data model, respectively. However, the ROS model has a significantly lower number of false negatives (4) compared to the original data model (56). This means that the ROS model is better at identifying high-risk loans than the original data model.
The classification reports for both models show that they both have very high precision scores for the healthy class. However, the precision score for the high-risk class is slightly lower for the ROS model (0.84) than the original data model (0.85). The recall score for the high-risk class is higher for the ROS model (0.99) compared to the original data model (0.91). The F1-score is also higher for the ROS model for the high-risk class (0.91) than the original data model (0.88).

Our balanced accuracy score and our accuracy score are both 99%. The balanced accuracy score is measures the performance of our model when the dataset is imbalanced. In our dataset however, we have an equal number in each class, 56271. Hence the balanced accuracy score is the same as the accuracy score.
Based on the confusion matrix, the model accurately accepted 18649 healthy loans and rejected 615 high-risk loans. However, the model also incorrectly rejected 116 healthy loans and incorrectly accepted 4 high-risk loans.
Our model's precision score is 84% for default loans and 100% for healthy loans. The model's recall score is 99% for default loans and 99% for healthy loans.
Our model's F1 score (the weighted average of precision and recall scores) is 91%.


Looking back at the original imbalanced data of 75036 instances of healthy loans compared tothe  2500 instances of high-risk laons, the resampled model improved the Balanced Accuracy score from ~0.952 to ~0.9937 and improved the recall for the high-risk loans from 91 % accuracy to 99% accuracy with a slighty decrease in precision for the high-risk laons, dropping from 85% accuracy to 84% accuracy. These results support a case for using oversampling to address the original imbalanced data.
Note: Oversampling, in the context of binary classification, generates additional synthetic samples for the minority class so that both classes have a more balanced representation in the dataset.


We can see that the logistic regression model fits well here too with an accuracy score of 99%. Getting into the details, we generally had ideal precision, recall, and f1 for the healthy loan to start with, and now we see those same percentages carry over with the oversampled data. The high-risk loan labels experienced a general increase in percent, except for the precision, which decreased by 1%. The recall and f1 score on the other hand experienced an 8% and 3% increase, respectively.



In [ ]:
#log_subroutine \
#    .EndProgramExecution()